crates/core/tedge_actors/src/test_helpers.rs

```diff
@@ -7,130 +7,307 @@ use crate::MessageSource;
+    /// # async fn main() -> Result<(),RuntimeError> {
+    ///
+    /// # use std::time::Duration;
+    /// let mut receiver_builder = SimpleMessageBoxBuilder::new("Recv", 16);
+    /// let sender_builder = SimpleMessageBoxBuilder::new("Send", 16).with_connection(&mut receiver_builder);
+    /// let mut sender = sender_builder.build();
+    /// let receiver: SimpleMessageBox<MyMessage,NoMessage> = receiver_builder.build();
+    ///
+    /// use tedge_actors::test_helpers::MessageReceiverExt;
+    /// let mut receiver = receiver.with_timeout(Duration::from_millis(100));
+    ///
+    /// sender.send(MyMessage::Foo(1)).await?;
+    /// sender.send(MyMessage::Bar(2)).await?;
+    /// sender.send(MyMessage::Foo(3)).await?;
+    ///
+    /// receiver.assert_received_unordered([
+    ///     MyMessage::Foo(3),
+    ///     MyMessage::Bar(2),
+    /// ]).await;
+    ///
+    /// # Ok(())
+    /// # }
+    ///
+    /// ```
+    async fn assert_received_unordered<Samples>(&mut self, expected: Samples)
+    where
+        Samples: IntoIterator + Send,
+        M: From<Samples::Item>;
+
+    /// Check that at least one matching message is received for each pattern.
+    ///
+    /// The messages can possibly be received in a different order or with interleaved messages.
+    ///
+    /// ```rust
+    /// use crate::tedge_actors::{Builder, NoMessage, RuntimeError, ServiceConsumer, SimpleMessageBox, SimpleMessageBoxBuilder, test_helpers};
+    ///
+    /// #[derive(Debug,Eq,PartialEq)]
+    /// enum MyMessage {
+    ///    Foo(u32),
+    ///    Bar(u32),
+    /// }
+    ///
+    /// impl MyMessage {
+    ///     pub fn count(&self) -> u32 {
+    ///         match self {
+    ///             MyMessage::Foo(n) => *n,
+    ///             MyMessage::Bar(n) => *n,
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(),RuntimeError> {
+    ///
+    /// use std::time::Duration;
+    /// let mut receiver_builder = SimpleMessageBoxBuilder::new("Recv", 16);
+    /// let sender_builder = SimpleMessageBoxBuilder::new("Send", 16).with_connection(&mut receiver_builder);
+    /// let mut sender = sender_builder.build();
+    /// let receiver: SimpleMessageBox<MyMessage,NoMessage> = receiver_builder.build();
+    ///
+    /// use tedge_actors::test_helpers::MessageReceiverExt;
+    /// let mut receiver = receiver.with_timeout(Duration::from_millis(100));
+    ///
+    /// sender.send(MyMessage::Foo(1)).await?;
+    /// sender.send(MyMessage::Bar(2)).await?;
+    /// sender.send(MyMessage::Foo(3)).await?;
+    ///
+    /// receiver.assert_received_matching(
+    ///     |pat:&u32,msg:&MyMessage| msg.count() == *pat,
+    ///     [3,2],
+    /// ).await;
+    ///
+    /// # Ok(())
+    /// # }
+    ///
+    /// ```
+    async fn assert_received_matching<T, F>(&mut self, matching: F, expected: T)
+    where
+        T: IntoIterator + Send,
+        F: Fn(&T::Item, &M) -> bool,
+        F: Send,
```

Comment:
> I'm assuming you didn't do it this way to avoid confusion on the return type of Fn that it is bool + Send

Exactly. The 2 lines version seems clearer to me.